### PR TITLE
Feature/2223 default types as schema org

### PR DESCRIPTION
**Pull request by kzangeli**
Fixes issue #2223

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -20,6 +20,7 @@
 - Fix: missing error payload for TooManyResults error in 'DELETE /v2/entities/<eId>' operation (Issue #1346)
 - Fix: correctly rending as JSON error responses that were being rendered as plain text (#1989, #1991) 
 - Fix: NGSIv2 "InvalidModification" errors changed to "Unprocessable" (according to rules defined in #1286)
+- Fix: Using default values for types of Entities, Attributes, and Metadata aligned with the type names used by schema.org (Issue #2223)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Add "... except for StructuredValue".

Hopefully, that will be solved before 1.3.0 get released (thus CNR entry adjusted) but in the case we fail at least the CNR entry will remain with a precise description.

Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in b310dd5

File: CHANGES_NEXT_RELEASE [hunk 2]
```diff
@@ -20,6 +20,7 @@
 - Fix: missing error payload for TooManyResults error in 'DELETE /v2/entities/<eId>' operation (Issue #1346)
 - Fix: correctly rending as JSON error responses that were being rendered as plain text (#1989, #1991) 
 - Fix: NGSIv2 "InvalidModification" errors changed to "Unprocessable" (according to rules defined in #1286)
+- Fix: Using default values for types of Entities, Attributes, and Metadata aligned with the type names used by schema.org, except for StructuredValue. (Issue #2223)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: After skype discussion, the last change in this CNR line should be rolled back.

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 3]
```diff
@@ -665,7 +665,7 @@ static void fillQueryEntity(BSONArrayBuilder& ba, const EntityId* enP)
     ent.append(idString, enP->id);
   }
 
-  if (enP->type != "")
+  if ((enP->type != "") && (enP->type != "Thing"))  // FIXME PR: Is this OK (Thing)? (POST /v2/op/query needs it ...)
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: I understand this approach makes "Thing" an alias of "any time". That is not the desired behaviour, i.e. a given user many query for entities which type is *exaclty* "Thing" and no other types.

Maybe defaulting should be applied by the upper layers (parsing, I guess) only in the case of "set" operations, i.e. create and update.

File: test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test [hunk 4]
```diff
@@ -381,4 +384,4 @@ Date: REGEX(.*)
 
 --TEARDOWN--
 brokerStop CB
-dbDrop CB
+#dbDrop CB
```
Comment (fgalan) [inline test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test]: Leftover

File: test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test [hunk 5]
```diff
@@ -30,6 +30,9 @@ brokerStart CB
 --SHELL--
 
 #
+# FIXME P6: steps 07, 09, 11, 13, and 15 in REGEXPECT of this test must change once the
```
Comment (fgalan) [inline test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test]: Probably we should close #2451 soon (I think it is quite clear no modification is needed, see my comment on that issue: https://github.com/telefonicaid/fiware-orion/issues/2451#issuecomment-242446354) thus removing these 3 lines before merging this PR.

File: test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_value_by_eid_fake_field_cases.test [hunk 6]
```diff
@@ -30,6 +30,9 @@ brokerStart CB
 --SHELL--
 
 #
+# FIXME P6: steps 04 and 06 in REGEXPECT of this test must change once the
```
Comment (fgalan) [inline test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_value_by_eid_fake_field_cases.test]: Probably we should close #2451 soon (I think it is quite clear no modification is needed, see my comment on that issue: https://github.com/telefonicaid/fiware-orion/issues/2451#issuecomment-242446354) thus removing these 3 lines before merging this PR.

File: test/functionalTest/cases/1068_metadata_value_as_compound/metadata_value_as_compound_array.test.DISABLED [hunk 7]
```diff
@@ -59,13 +61,21 @@ echo
 
 echo "02. GET /v2/entities, to see the entity"
 echo "======================================="
+payload='{
```
Comment (fgalan) [inline test/functionalTest/cases/1068_metadata_value_as_compound/metadata_value_as_compound_array.test.DISABLED]: GET operation doesn't need payload...

File: test/functionalTest/cases/1188_empty_attribute_value/empty_attribute_value.test [hunk 8]
```diff
@@ -30,12 +30,17 @@ brokerStart CB
 --SHELL--
 
 #
+# FIXME P6: step 06 in REGEXPECT of this test must change once the
```
Comment (fgalan) [inline test/functionalTest/cases/1188_empty_attribute_value/empty_attribute_value.test]: Probably we should close #2451 soon (I think it is quite clear no modification is needed, see my comment on that issue: https://github.com/telefonicaid/fiware-orion/issues/2451#issuecomment-242446354) thus removing these 3 lines before merging this PR.

File: test/functionalTest/cases/1259_new_json_for_v2/put_attribute_value.test [hunk 9]
```diff
@@ -30,6 +30,9 @@ brokerStart CB 0-255
 --SHELL--
 
 #
+# FIXME P6: steps 04, 06, 08, 10, 12, 14, 16 18, and 20 in REGEXPECT of this test must change once the
```
Comment (fgalan) [inline test/functionalTest/cases/1259_new_json_for_v2/put_attribute_value.test]: Probably we should close #2451 soon (I think it is quite clear no modification is needed, see my comment on that issue: https://github.com/telefonicaid/fiware-orion/issues/2451#issuecomment-242446354) thus removing these 3 lines before merging this PR.

File: test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    "type": "Thing"
+}
+
+
+--TEARDOWN--
+brokerStop CB
+dbDrop CB
+
+
+#11. Create E6/A1/M1, no type, value of A1/M1 is a compound object
```
Comment (fgalan) [inline test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test]: Leftover maybe already moved (or pending to move) to a separate .test in this directory?

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: (BTW maybe this change is related with the mysterious .DISABLED .test in this PR)

File: test/functionalTest/cases/1068_metadata_value_as_compound/metadata_value_as_compound_array.test.DISABLED (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/cases/1068_metadata_value_as_compound/metadata_value_as_compound_array.test.DISABLED]: Yeah :-)
Not sure how this got in there, probably a cut 'n paste error ... 
The commented option --noPayloadCheck will be removed as well.

File: test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test]: The entire test is funny ... starting with 09, clearly a copy-paste from some other test and then the last two steps commented.
It seems like something I just for debugging only and it got 'git-added' by mistake.

I will take a closer look and remove it if so, or fix it if not.

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Redone in a8eb742

File: CHANGES_NEXT_RELEASE (hunk 2 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in a8eb742

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Fixed in a8eb742

More details about the mysterious .DISABLED test, pls

File: test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test]: Fixed in a8eb742

File: test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test]: Fixed in a8eb742

File: test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_value_by_eid_fake_field_cases.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/0999_update_single_attr_value_by_eid/update_single_attr_value_by_eid_fake_field_cases.test]: Fixed in a8eb742

File: test/functionalTest/cases/1068_metadata_value_as_compound/metadata_value_as_compound_array.test.DISABLED (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/cases/1068_metadata_value_as_compound/metadata_value_as_compound_array.test.DISABLED]: Fixed in a8eb742

File: test/functionalTest/cases/1188_empty_attribute_value/empty_attribute_value.test (hunk 8 above)
Comment (kzangeli) [inline test/functionalTest/cases/1188_empty_attribute_value/empty_attribute_value.test]: Fixed in a8eb742

File: test/functionalTest/cases/1259_new_json_for_v2/put_attribute_value.test (hunk 9 above)
Comment (kzangeli) [inline test/functionalTest/cases/1259_new_json_for_v2/put_attribute_value.test]: Fixed in a8eb742

File: test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+Default Metadata Types as schema.org
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 09. Create E5/A1/M1, no type, value of A1/M1 is a compound vector
```
Comment (kzangeli) [inline test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test]: Fixed in a8eb742 - As suspected, this test was just a temporary debugging tool. Removed from PR

File: test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/2223_default_types_schema_org/default_metadata_types_array_schema_org.test]: Fixed in a8eb742 - As suspected, this test was just a temporary debugging tool. Removed from PR

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: By mysterious .DISABLED test I mean 1068_metadata_value_as_compound/metadata_value_as_compound_array.test.DISABLED, to be "recovered" just after this PR.

**Comment by fgalan** [comment]
LGTM

**Pull request merged**